packages/block-editor/src/store/selectors.js

```diff
@@ -2722,11 +2722,13 @@ export function isBlockVisible( state, clientId ) {
  */
 export const __unstableGetVisibleBlocks = createSelector(
 	( state ) => {
-		return new Set(
-			Object.keys( state.blockVisibility ).filter(
-				( key ) => state.blockVisibility[ key ]
-			)
+		const visibleBlocks = Object.keys( state.blockVisibility ).filter(
+			( key ) => state.blockVisibility[ key ]
 		);
+		if ( visibleBlocks.length === 0 ) {
+			return EMPTY_ARRAY;
+		}
+		return visibleBlocks;
```

Comment:
Why change the type from `Set`? This should break things and I'm surprized the CI is green. We could have an `EMPTY_SET` I guess, no?

Reply:
Changed back to a `Set` in https://github.com/WordPress/gutenberg/pull/47681/commits/bc5e01f3625c260b65e91392cf721c2bccfa0cdd.

Reply:
Same here. That, plus the fact that it's marked as unstable is why I was considering simplifying it to an array. But I don't have such a strong feeling about whether to go with a `Set` or an `Array`, as long as the optimization is there.

Reply:
Good point, I think that's the reason 👍 

Maybe ideally this problem could have been solved in the `blockVisibility` reducer, but I guess that can be done another time.